From 90587d4ab9625460bf89ebcacdc8133c688e3900 Mon Sep 17 00:00:00 2001
From: Edgar E. Iglesias <edgar.iglesias@gmail.com>
Date: Sun, 21 Aug 2011 22:01:29 +0200
Subject: [PATCH 02/79] gcc-microblaze: Bailout early for PIC modes when target does not do PIC.

2011-08-21  Edgar E. Iglesias  <edgar.iglesias@gmail.com>

	* config/microblaze/linux.h (TARGET_SUPPORTS_PIC): Define as 1.
	* config/microblaze/microblaze.h (TARGET_SUPPORTS_PIC): Define
          as 0.
	* gcc/config/microblaze/microblaze.c (microblaze_option_override):
          Bail out early for PIC modes when target does not support PIC.

Signed-off-by: Edgar E. Iglesias <edgar.iglesias@gmail.com>
---
 gcc/config/microblaze/linux.h      |    2 ++
 gcc/config/microblaze/microblaze.c |   10 ++++++++++
 gcc/config/microblaze/microblaze.h |    3 +++
 3 files changed, 15 insertions(+), 0 deletions(-)

diff --git a/gcc/config/microblaze/linux.h b/gcc/config/microblaze/linux.h
index 3c13fab..8b59d29 100644
--- a/gcc/config/microblaze/linux.h
+++ b/gcc/config/microblaze/linux.h
@@ -19,6 +19,8 @@
    along with GCC; see the file COPYING3.  If not see
    <http://www.gnu.org/licenses/>.  */
 
+#undef TARGET_SUPPORTS_PIC
+#define TARGET_SUPPORTS_PIC 1
 
 #define DYNAMIC_LINKER "/lib/ld.so.1"
 #undef  SUBTARGET_EXTRA_SPECS
diff --git a/gcc/config/microblaze/microblaze.c b/gcc/config/microblaze/microblaze.c
index b50c794..e3c7294 100644
--- a/gcc/config/microblaze/microblaze.c
+++ b/gcc/config/microblaze/microblaze.c
@@ -1302,6 +1302,16 @@ microblaze_option_override (void)
 				  ? g_switch_value
 				  : MICROBLAZE_DEFAULT_GVALUE);
 
+  if (flag_pic)
+    {
+      if (!TARGET_SUPPORTS_PIC)
+        {
+          error ("-fPIC/-fpic not supported for this target");
+          /* Clear it to avoid further errors.  */
+          flag_pic = 0;
+        }
+    }
+
   /* Check the MicroBlaze CPU version for any special action to be done.  */
   if (microblaze_select_cpu == NULL)
     microblaze_select_cpu = MICROBLAZE_DEFAULT_CPU;
diff --git a/gcc/config/microblaze/microblaze.h b/gcc/config/microblaze/microblaze.h
index f60ab6b..8748b17 100644
--- a/gcc/config/microblaze/microblaze.h
+++ b/gcc/config/microblaze/microblaze.h
@@ -49,6 +49,9 @@ extern enum pipeline_type microblaze_pipe;
 /* Default target_flags if no switches are specified  */
 #define TARGET_DEFAULT      (MASK_SOFT_MUL | MASK_SOFT_DIV | MASK_SOFT_FLOAT)
 
+/* The default is to not support PIC.  */
+#define TARGET_SUPPORTS_PIC 0
+
 /* What is the default setting for -mcpu= . We set it to v4.00.a even though 
    we are actually ahead. This is safest version that has generate code 
    compatible for the original ISA */
-- 
1.7.1

